### PR TITLE
Avoid logits multimem all-gather on cross-node TP groups

### PR DESCRIPTION
**Pull request by san-tian**
## Summary

- disable the logits `MultimemAllGatherer` fast path when the TP process group spans multiple nodes
- fall back to the existing normal tensor-parallel all-gather path before calling `torch.distributed._symmetric_memory.rendezvous`
- add unit coverage for cross-node vs single-node TP group behavior

## Why

We hit a startup hang on a multi-node H20 deployment where logits multimem all-gather attempted to create symmetric memory over a cross-node TP group during target-verify CUDA graph capture. This path failed while establishing the symmetric-memory handle and left the server unable to become ready. SGLang already disables other custom all-reduce fast paths for process groups spanning nodes; this applies the same guard to logits multimem all-gather.

## Reproduction environment

Hardware/topology:

- 4 H20 hosts, each 8 x NVIDIA H20-3e
- TP16: 2 hosts x 8 GPUs
- TP32: 4 hosts x 8 GPUs
- tensor-parallel groups span multiple hosts

Software:

- PyTorch `2.11.0+cu130`
- CUDA `13.0`
- SGLang source based on a recent `main`/deployment branch containing GLM-5.2 support

Models:

- GLM-5.2 FP8
- GLM-5.2 BF16

FP8 TP16 launch shape:

```bash
python3 -m sglang.launch_server \
  --model-path <GLM-5.2-FP8-path> \
  --served-model-name <served-model-name> \
  --host 0.0.0.0 --port 30000 \
  --tp-size 16 --nnodes 2 --node-rank <0|1> \
  --dist-init-addr <rank0-host>:<port> \
  --context-length 1048576 \
  --speculative-algorithm EAGLE \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --mem-fraction-static 0.80 \
  --max-running-requests 16 \
  --cuda-graph-max-bs-decode 16 \
  --enable-metrics --enable-mfu-metrics \
  --disable-custom-all-reduce \
  --disable-overlap-schedule
```

BF16 TP32 launch shape:

```bash
python3 -m sglang.launch_server \
  --model-path <GLM-5.2-BF16-path> \
  --served-model-name <served-model-name> \
  --host 0.0.0.0 --port 30000 \
  --tp-size 32 --nnodes 4 --node-rank <0|1|2|3> \
  --dist-init-addr <rank0-host>:<port> \
  --context-length 1048576 \
  --speculative-algorithm EAGLE \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --mem-fraction-static 0.75 \
  --max-running-requests 16 \
  --cuda-graph-max-bs-decode 32 \
  --enable-metrics --enable-mfu-metrics \
  --disable-custom-all-reduce \
  --disable-overlap-schedule
```

Observed logs:

```text
Capture target verify CUDA graph begin. backend=full, num_tokens_per_bs=4, ...
multimem all-gather disabled (Failed to send fd: No such file or directory)
```

A `py-spy` snapshot showed the blocked path under:

```text
logits_processor.py:_get_logits
triton_symm_mem_ag.MultimemAllGatherer
torch.distributed._symmetric_memory.rendezvous
```

Temporary workaround used during deployment:

```bash
export SGLANG_DISABLE_LOGITS_MULTIMEM_AG=1
```

This avoids constructing the logits multimem all-gather path and lets SGLang use the existing normal tensor-parallel all-gather fallback. It was used as a deployment workaround before this PR, together with CUDA graph and MTP still enabled.

With a local patch that disabled only logits multimem all-gather, the same deployment became ready with CUDA graph and MTP still enabled. A validated 1M FP8 TP16 service used:

```bash
--max-total-tokens 1048576 --kv-cache-dtype fp8_e4m3
```

and reported:

```text
context_length=1048576
max_total_tokens=1048576
max_total_num_tokens=1048576
kv_cache_dtype=fp8_e4m3
speculative_algorithm=EAGLE
disable_cuda_graph=False
disable_decode_cuda_graph=False
```

## Tests

```bash
PYTHONPATH=python python3 -m pytest test/srt/test_triton_symm_mem_ag.py -q
python3 -m py_compile python/sglang/srt/distributed/device_communicators/triton_symm_mem_ag.py test/srt/test_triton_symm_mem_ag.py
git diff --check
```

<sub>✨ Presented to you with <a href=" ">Mind Lab</a > — A Lab for Experiential Intelligence.</sub>

---
### CI States

Latest PR Test (Base): :x: [Run #28634147034](https://github.com/sgl-project/sglang/actions/runs/28634147034)
Latest PR Test (Extra): :x: [Run #28689676978](https://github.com/sgl-project/sglang/actions/runs/28689676978)

### CODE REVIEW

File: test/srt/test_triton_symm_mem_ag.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+from types import SimpleNamespace
```
Comment (Fridge003) [inline test/srt/test_triton_symm_mem_ag.py]: No need to add this test

File: python/sglang/srt/distributed/device_communicators/triton_symm_mem_ag.py
Comment (Fridge003) [inline python/sglang/srt/distributed/device_communicators/triton_symm_mem_ag.py:464]: Can we move the logics here.
Make self._state None when detecting multi nodes

**Review by Fridge003 — CHANGES_REQUESTED** [review]

File: python/sglang/srt/distributed/device_communicators/triton_symm_mem_ag.py
Comment (san-tian) [reply python/sglang/srt/distributed/device_communicators/triton_symm_mem_ag.py:464]: Updated. I removed the test and moved the cross-node TP check into `MultimemAllGatherer` initialization.

**Review by Fridge003 — APPROVED** [review]

**Review by mmangkad — CHANGES_REQUESTED** [review]
We can make this simpler and more correct by moving the check into `_build()` before `create_state()`, so it guards the actual rendezvous group instead of the constructor-time group

**Comment by Fridge003** [comment]
> We can make this simpler and more correct by moving the check into `_build()` before `create_state()`, so it guards the actual rendezvous group instead of the constructor-time group 

In this case it should be OK, since the rendezvous group is the same as constructor-time group at its creation

**Comment by mmangkad** [comment]
> > We can make this simpler and more correct by moving the check into `_build()` before `create_state()`, so it guards the actual rendezvous group instead of the constructor-time group
> 
> In this case it should be OK, since the rendezvous group is the same as constructor-time group at its creation

Hmm, I was thinking `_build()` made more sense since it re-reads the TP group right before rendezvous, but sure

**Review by mmangkad — COMMENTED** [review]
Dismiss review

**Review by mmangkad — APPROVED** [review]
Dismiss review

**Review by Fridge003 — APPROVED** [review]
